flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/DynamicKafkaSource.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.kafka.source;
```

Comment:
On hindsight, I wonder whether `source.dynamic` makes more sense than `dynamic.source` 🙈

Reply:
I actually prefer `dynamic.source` since it makes sense for future usecases to leverage the dynamic code. e.g. a `dynamic.sink`. With that being said, I'm going to move the metadata up one package since that isn't source specific.

Reply:
Let me know if you have any other considerations. I'll squash the commits if we are in agreement (leaving the dangling commit now for your reviewing convenience)